velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -202,310 +195,218 @@ inline static TResult addLarge(
+// addition.
+template <typename TResult, typename A, typename B>
+inline void applyAdd(
+    TResult& r,
+    A a,
+    B b,
+    uint8_t aRescale,
+    uint8_t bRescale,
+    uint8_t aScale,
+    uint8_t bScale,
+    uint8_t rPrecision,
+    uint8_t rScale,
+    bool& overflow) {
+  if (rPrecision < LongDecimalType::kMaxPrecision) {
+    const int128_t aRescaled = a * velox::DecimalUtil::kPowersOfTen[aRescale];
+    const int128_t bRescaled = b * velox::DecimalUtil::kPowersOfTen[bRescale];
+    r = TResult(aRescaled + bRescaled);
+  } else {
+    const uint32_t minLeadingZeros =
+        sparksql::DecimalUtil::minLeadingZeros<A, B>(a, b, aRescale, bRescale);
+    if (minLeadingZeros >= 3) {
+      // Fast path for no overflow. If both numbers contain at least 3 leading
+      // zeros, they can be added directly without the risk of overflow.
+      // The reason is if a number contains at least 2 leading zeros, it is
+      // ensured that the number fits in the maximum of decimal, because
+      // '2^126 - 1 < 10^38 - 1'. If both numbers contain at least 3 leading
+      // zeros, we are guaranteed that the result will have at least 2 leading
+      // zeros.
+      int128_t aRescaled = a * velox::DecimalUtil::kPowersOfTen[aRescale];
+      int128_t bRescaled = b * velox::DecimalUtil::kPowersOfTen[bRescale];
+      r = reduceScale(
+          TResult(aRescaled + bRescaled), std::max(aScale, bScale) - rScale);
     } else {
-      // Fast path if one or more arguments are encoded.
-      exec::DecodedArgs decodedArgs(rows, args, context);
-      auto a = decodedArgs.at(0);
-      auto b = decodedArgs.at(1);
-      context.applyToSelectedNoThrow(rows, [&](auto row) {
-        bool overflow = false;
-        Operation::template apply<R, A, B>(
-            rawResults[row],
-            a->valueAt<A>(row),
-            b->valueAt<B>(row),
-            aRescale_,
-            bRescale_,
-            aPrecision_,
-            aScale_,
-            bPrecision_,
-            bScale_,
-            rPrecision_,
-            rScale_,
-            overflow);
-        if (overflow ||
-            !velox::DecimalUtil::valueInPrecisionRange(
-                rawResults[row], rPrecision_)) {
-          result->setNull(row, true);
-        }
-      });
+      // The risk of overflow should be considered. Add whole and fraction
+      // parts separately, and then combine.
+      r = addLarge<TResult, A, B>(a, b, aScale, bScale, rScale, overflow);
     }
   }
+}
 
- private:
-  R* prepareResults(
-      const SelectivityVector& rows,
-      const TypePtr& resultType,
-      exec::EvalCtx& context,
-      VectorPtr& result) const {
-    context.ensureWritable(rows, resultType, result);
-    result->clearNulls(rows);
-    return result->asUnchecked<FlatVector<R>>()->mutableRawValues();
-  }
+// Compute the add and subtract operation result precision and scale following
+// Hive's formulas.
```

Comment:
Comment suggestion:

> // Computes the result precision and scale for decimal add and subtract operations following
// Hive's formulas. If result is representable with long decimal, the result
// scale is the maximum of 'aScale' and 'bScale'. If not, reduces result scale
// and caps the result precision at 38.